rclcpp/test/rclcpp/test_qos_event.cpp

```diff
@@ -561,6 +618,11 @@ TEST_F(TestQosEvent, test_pub_matched_event_by_option_event_callback)
 
 TEST_F(TestQosEvent, test_sub_matched_event_by_option_event_callback)
 {
+  std::string rmw_implementation_str = std::string(rmw_get_implementation_identifier());
```

Comment:
matched events should definitely work else there is a bug in rmw_zenoh.

Reply:
failing here too 

```
Expected equality of these values:
  s.total_count
    Which is: 7
  matched_expected_result.total_count
    Which is: 1

/home/ahcorde/ros2_rolling/src/ros2/rclcpp/rclcpp/test/rclcpp/test_qos_event.cpp:620: Failure
Expected equality of these values:
  s.total_count_change
    Which is: 2
  matched_expected_result.total_count_change
    Which is: 1

/home/ahcorde/ros2_rolling/src/ros2/rclcpp/rclcpp/test/rclcpp/test_qos_event.cpp:621: Failure
Expected equality of these values:
  s.current_count
    Which is: 3
  matched_expected_result.current_count
    Which is: 1

/home/ahcorde/ros2_rolling/src/ros2/rclcpp/rclcpp/test/rclcpp/test_qos_event.cpp:619: Failure
Expected equality of these values:
  s.total_count
    Which is: 8
  matched_expected_result.total_count
    Which is: 1

/home/ahcorde/ros2_rolling/src/ros2/rclcpp/rclcpp/test/rclcpp/test_qos_event.cpp:620: Failure
Expected equality of these values:
  s.total_count_change
    Which is: 1
  matched_expected_result.total_count_change
    Which is: 0

/home/ahcorde/ros2_rolling/src/ros2/rclcpp/rclcpp/test/rclcpp/test_qos_event.cpp:621: Failure
Expected equality of these values:
  s.current_count
    Which is: 2
  matched_expected_result.current_count
    Which is: 0

```

Reply:
So this relates to what I mentioned above where we should update the test similar to 4) in https://github.com/ros2/rcl/pull/1164. Basically more QoS combinations are compatible in rmw_zenoh so we should set the `matched_expected_result.current_count` to the output of `rmw_qos_profile_check_compatible` checks.

Reply:
There might be some tests that we can fix with your comment, but the subscription counter is wrong, it's only increasing.

If you comment out some tests the numbers are different which somehow don't clean the counters between tests.

Reply:
This looks like a bug in `rmw_zenoh`. I would suggest we don't skip this test and instead have it fail so that we can address it.

We should only skip the tests for functionalities we do not support.

Reply:
After looking at the problem closer; the issue is that 

1. rmw_init() is called once in `SetUpTestCase()` which creates the Zenoh session and initializes the ROS graph.
2. Each time a new test case runs, we run `SetUp()` and `TearDown()` which creates and destroys the node and some publishes and subscriptions. However, since the session is still the same throughout the tests each time a new pub/sub match is found, the same `total_count` and `total_count_change` updates. From the sessions's perspective, the total count is indeed valid since it was never shutdown.

In https://github.com/ros2/rclcpp/pull/2626/commits/6c9babacc1b68f43861906752e4e3f974ced3c5d, I moved the `rmw_init()` command inside `SetUp()` and the `rmw_shutdown()` inside` TearDown()`, the total_count numbers are accurate. With https://github.com/ros2/rmw_zenoh/pull/287 all events tests are passing.